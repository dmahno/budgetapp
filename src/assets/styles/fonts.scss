@font-face {
  font-family: 'Golos';
  src:
    url('../fonts/Golos-Regular.woff2') format('woff2'),
    url('../fonts/fonts/Golos-Regular.ttf') format('truetype');
  font-weight: 400;
  font-style: normal;
}

@font-face {
  font-family: 'Golos';
  src:
    url('../fonts/fonts/Golos-Medium.woff2') format('woff2'),
    url('../fonts/fonts/Golos-Medium.ttf') format('truetype');
  font-weight: 500;
  font-style: normal;
}

@font-face {
  font-family: 'Golos';
  src:
    url('../fonts/fonts/Golos-SemiBold.woff2') format('woff2'),
    url('../fonts/fonts/Golos-SemiBold.ttf') format('truetype');
  font-weight: 600;
  font-style: normal;
}

@font-face {
  font-family: 'Golos';
  src:
    url('../fonts/fonts/Golos-Bold.woff2') format('woff2'),
    url('../fonts/fonts/Golos-Bold.ttf') format('truetype');
  font-weight: 700;
  font-style: normal;
}

@font-face {
  font-family: 'Golos';
  src:
    url('../fonts/fonts/GolosText-Black.woff2') format('woff2'),
    url('../fonts/fonts/GolosText-Black.ttf') format('truetype');
  font-weight: 800;
  font-style: normal;
}
