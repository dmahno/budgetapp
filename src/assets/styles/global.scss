@use 'assets/styles/fonts';
@use 'assets/styles/mobile';
@use 'assets/variables/colors';

* {
  box-sizing: border-box;
}

html,
body {
  margin: 0;
  padding: 0;
  size: 16px;
  font-family: 'Golos Text', serif;
  background: var(--backgroundColor);
  color: var(--textColor);
}

body {
  min-width: 320px;
  padding: 0 $indent-s;
}

@include mobile {
  body {
    font-size: 14px;
  }
}

h1,
h2,
h3,
h4,
h5,
h6,
p,
input,
button,
textarea,
ul,
li {
  margin: 0;
  padding: 0;
}

li {
  list-style-type: none;
}

.link,
.active {
  transition:
    background-color 0.2s ease,
    color 0.2s ease;
}

svg {
  fill: var(--iconColor);
}

a {
  color: var(--linkColor);
}

input:-webkit-autofill,
input:-webkit-autofill:focus {
  transition:
    background-color 0s 600000s,
    color 0s 600000s !important;
}

button {
  background-color: var(--buttonBackgroundColor);
  color: var(--buttonTextColor);
  border: none;
  padding: 10px 20px;
  cursor: pointer;
}

html {
  scroll-behavior: smooth;
}
