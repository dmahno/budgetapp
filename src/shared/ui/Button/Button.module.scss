.button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: 16px;
  font-size: 16px;
  cursor: pointer;
  width: auto;
  transition: all 0.3s ease;

  &.loading {
    cursor: progress;
  }

  &.disabled {
    cursor: not-allowed;
    opacity: 0.6;
  }

  &.fullWidth {
    width: 100%;
  }
}

.size-s {
  font-size: 12px;
  padding: 6px 12px;
}

.size-m {
  font-size: 16px;
  height: 56px;
  padding: 8px 16px;
}

.size-l {
  font-size: 20px;
  padding: 10px 20px;
}

.appearance-primary {
  background-color: var(--buttonPrimaryBackgroundColor);
  color: var(--buttonPrimaryTextcolor);

  &:hover {
    background-color: var(--buttonPrimaryBackgroundColorHover);
  }
}

.appearance-secondary {
  background-color: var(--secondary-color);
  color: #fff;

  &:hover {
    background-color: var(--secondary-hover);
  }
}

.appearance-thirdly {
  background-color: var(--thirdly-color);
  color: #fff;

  &:hover {
    background-color: var(--thirdly-hover);
  }
}

.icon {
  display: inline-flex;
  margin: 0 4px;
}

.loader {
  border: 2px solid rgba(255, 255, 255, 0.3);
  border-radius: 50%;
  border-top: 2px solid #fff;
  width: 16px;
  height: 16px;
  animation: spin 0.8s linear infinite;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
