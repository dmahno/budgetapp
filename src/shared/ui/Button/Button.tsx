import React, {ReactNode, MouseEventHandler, ElementType} from 'react';
import classNames from 'classnames';

import {Spinner} from 'shared/ui';

import styles from './Button.module.scss';

type TButtonSizes = 's' | 'm' | 'l';
type TButtonAppearances = 'primary' | 'secondary' | 'thirdly';
type TButtonWidths = 'auto' | 'full' | string;

interface IButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  id?: string;
  name?: string;
  children?: ReactNode | ReactNode[];
  disabled?: boolean;
  as?: ElementType;
  className?: string;
  loading?: boolean;
  icon?: JSX.Element;
  iconAlign?: 'left' | 'right';
  onClick?: MouseEventHandler<HTMLButtonElement>;
  size?: TButtonSizes;
  appearance?: TButtonAppearances;
  width?: TButtonWidths;
}

export const Button: React.FC<IButtonProps> = ({
  id,
  name,
  children,
  disabled = false,
  as: Component = 'button',
  className,
  loading = false,
  icon,
  iconAlign = 'left',
  onClick,
  size = 'm',
  appearance = 'primary',
  width = 'auto',
  ...props
}) => {
  const buttonClass = classNames(
    styles.button,
    styles[`size-${size}`],
    styles[`appearance-${appearance}`],
    {
      [styles.loading]: loading,
      [styles.disabled]: disabled,
      [styles.fullWidth]: width === 'full',
    },
    className,
  );

  const buttonStyle =
    width !== 'full' && width !== 'auto' ? {width} : undefined;

  return (
    <Component
      id={id}
      name={name}
      className={buttonClass}
      onClick={onClick as MouseEventHandler<HTMLButtonElement>}
      disabled={disabled}
      style={buttonStyle}
      {...props}
    >
      {loading ? (
        <Spinner size="s" mode="dark" data-testid="spinner" />
      ) : (
        <>
          {icon && iconAlign === 'left' && (
            <span data-testid="icon" className={styles.icon}>
              {icon}
            </span>
          )}
          {children}
          {icon && iconAlign === 'right' && (
            <span data-testid="icon" className={styles.icon}>
              {icon}
            </span>
          )}
        </>
      )}
    </Component>
  );
};
