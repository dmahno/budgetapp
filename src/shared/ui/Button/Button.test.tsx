import {render, screen, fireEvent} from '@testing-library/react';
import {describe, it, expect, vi} from 'vitest';

import LogoImage from 'shared/assets/logo.svg';

import {Button} from './Button';
import styles from './Button.module.scss';

describe('Button', () => {
  it('renders with default props', () => {
    const {container} = render(<Button>Click Me</Button>);
    const button = screen.getByRole('button');

    expect(button).toHaveClass(
      styles.button,
      styles['size-m'],
      styles['appearance-primary'],
    );
    expect(button).toHaveTextContent('Click Me');
    expect(container).toMatchSnapshot();
  });

  it('renders with a custom size and appearance', () => {
    render(
      <Button size="l" appearance="secondary">
        Custom Button
      </Button>,
    );
    const button = screen.getByRole('button');

    expect(button).toHaveClass(
      styles.button,
      styles['size-l'],
      styles['appearance-secondary'],
    );
    expect(button).toHaveTextContent('Custom Button');
  });

  it('renders with full width', () => {
    render(<Button width="full">Full Width Button</Button>);
    const button = screen.getByRole('button');

    expect(button).toHaveClass(styles.fullWidth);
  });

  it('applies custom width style', () => {
    render(<Button width="200px">Custom Width Button</Button>);
    const button = screen.getByRole('button');

    expect(button).toHaveStyle({width: '200px'});
  });

  it('renders with a loading spinner when loading is true', () => {
    render(<Button loading>Loading Button</Button>);
    const spinner = screen.getByTestId('spinner');

    expect(spinner).toBeInTheDocument();
  });

  it('renders with an icon aligned to the left', () => {
    render(
      <Button icon={<LogoImage />} iconAlign="left">
        Button with Icon
      </Button>,
    );

    const iconElement = screen.getByTestId('icon');
    expect(iconElement).toBeInTheDocument();
    expect(iconElement.nextSibling).toHaveTextContent('Button with Icon');
  });

  it('renders with an icon aligned to the right', () => {
    render(
      <Button icon={<LogoImage />} iconAlign="right">
        Button with Icon
      </Button>,
    );

    const iconElement = screen.getByTestId('icon');
    expect(iconElement).toBeInTheDocument();
    expect(iconElement.previousSibling).toHaveTextContent('Button with Icon');
  });

  it('triggers onClick handler when clicked', () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Clickable Button</Button>);
    const button = screen.getByRole('button');

    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('renders with disabled style and prevents clicks when disabled', () => {
    const handleClick = vi.fn();
    render(
      <Button onClick={handleClick} disabled>
        Disabled Button
      </Button>,
    );
    const button = screen.getByRole('button');

    expect(button).toHaveClass(styles.disabled);
    expect(button).toBeDisabled();

    fireEvent.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });
});
