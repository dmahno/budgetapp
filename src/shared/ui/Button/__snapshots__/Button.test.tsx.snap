// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders with default props 1`] = `
<div>
  <button
    class="_button_0302f3 _size-m_0302f3 _appearance-primary_0302f3"
  >
    Click Me
  </button>
</div>
`;
